File: erlang/seminars/dns/src/dns.erl
-module(dns).

%%% https://www.ietf.org/rfc/rfc1035.txt

-compile(export_all).

-define(Server, "8.8.8.8").  %% The Google DNS server

-define(Port, 53).

-define(Local, 5300).

-define(Timeout, 4000).

start() ->
    start(?Local, ?Server).

start(Port) ->
    start(Port, ?Server).    

start(Port, Server) -> 
    spawn_link(fun() -> init(Port, Server) end).

init(Port, Server) ->
    case gen_udp:open(Port, [{active, true}, binary]) of
	{ok, Socket} ->
	    dns(Socket, Server);
	Error ->
	    io:format("dns error opening server socket ~w~n", [Error])
    end.
    


dns(Socket, Server) ->
    receive 
	{udp, Socket, IP, Port, Packet} ->
	    Reply = fun(Rep) -> gen_udp:send(Socket, IP, Port, Rep)  end,
	    Frw = fun(Req) -> forward(Req, Server) end,
 	    handler:start(Packet, Reply, Frw),

	    dns(Socket, Server);
	update ->
	    dns:dns(Socket, Server);
	stop ->
	    io:format("by by~n", []),
	    ok;
	Error ->
	    io:format("strange message ~w~n", [Error]),
	    dns(Socket, Server)
    end.






%         Reply = fun(Rep) -> reply(Rep, Socket, IP, Port) 
%          Frw = fun(Req) -> forward(Req, Server) end,


% 	    handler:handle(Packet, Reply, Frw),

%	    handler:start(Packet, Reply, Frw),

%   	    handler:handle(Packet, IP, Port, Socket, Server),





forward(Request, Server) ->
    case gen_udp:open(0, [{active, true}, binary]) of
	{ok, Client} ->
	    gen_udp:send(Client, Server, ?Port, Request),
	    Result = receive 
			 {udp, Client, _IP, _Port, Reply} ->
			     {ok, Reply}
		     after ?Timeout ->
			     {error, timeout}
		     end,
	    gen_udp:close(Client),
	    Result;
	Error ->
	    {error, Error}
    end.
    

File: exercises/dns/src/dns.ex
defmodule DNS do

  @server {8,8,8,8}
  @port 53
  @local 5300
  
  def start() do
    start(@local, @server, @port)
  end

  def start(local, server, port) do
    spawn(fn() -> init(local, server, port) end)
  end

  def init(local, server, port) do
    case :gen_udp.open(local, [{:active, true}, :binary]) do
      {:ok, local} ->
	case :gen_udp.open(0, [{:active, true}, :binary]) do
	  {:ok, remote} ->
            dns(local, remote, server, port, 0, Map.new())
	  error ->
            :io.format("DNS error opening remote socket: ~w~n", [error])
	end
      error ->
        :io.format("DNS error opening local socket: ~w~n", [error])
    end
  end

  def dns(local, remote, server, server_port, n, queries) do

    receive do
      {:udp, ^local, client, client_port, msg} ->
	:io.format("message from  ~w:~w~n", [client, client_port])
	{id, _qr, _op, _aa, _tc, _rd, _ra, _resp, body} = Msg.decode(msg)
	:io.format("id: ~w, body: ~p~n", [id, body])
	query = Msg.fake(msg, n)
	:gen_udp.send(remote, server, server_port, query)
	dns(local, remote, server, server_port, n+1, Map.put(queries, n, {id, client, client_port}))

      {:udp, ^remote, server, server_port, msg} ->
	:io.format("reply from  ~w:~w~n", [server, server_port])
	{id, _qr, _op, _aa, _tc, _rd, _ra, _resp, reply} = Msg.decode(msg)
	:io.format("id: ~w, reply: ~p~n", [id, reply])
	{{id, client, client_port}, rest} = Map.pop(queries, id)
	response = Msg.fake(msg, id)
	:gen_udp.send(local, client, client_port, response)
	dns(local, remote, server, server_port, n, rest)	

      :update ->
	DNS.dns(local, remote, server, server_port, n, queries)

      :status ->
	:io.format("local: ~w~nremote: ~w~nserver ~w~nport ~w~n", [local, remote, server, server_port])
	dns(local, remote, server, server_port, n, queries)

      :quit ->
	:ok

      strange ->
	:io.format("strange: ~w~n", [strange])
	dns(local, remote, server, server_port, n, queries)
    end
  end    

  
end
